### ISSUE TITLE
Directive#postLink and Directive#withController

### ISSUE THREAD

**Issue by jpablo** [jokade/scalajs-angulate#28]
First I must say that I'm excited about this project, it is already usable and I like the fact that the scala api it is very similar to the js api.

I have a few questions (sorry for the long post):

Maybe I'm not using correctly `withController`, but at first I thought it is similar to using angular's native controller from the directive api.

In that case, the fourth argument to the `postLink` method should be an instance of the associated controller. For example, in your angulate-todomvc Directives.scala, the fourth argument (`controller: js.Dynamic`) of `postLink`, should be an instance of `TodoItemDirective.Ctrl`.

``` scala
class TodoItemDirective extends Directive {
  override val templateUrl = "/web/todo-item.html"

  override def isolateScope: Dictionary[String] = js.Dictionary(
    "todo" -> "=item",
    "fireOnRemove" -> "&onRemove",
    "fireOnChange" -> "&onChange"
  )

  override type withController = TodoItemDirective.Ctrl

  // The controller argument should be an instance of TodoItemDirective.Ctrl
  override def postLink(scope: Scope, elem: JQLite, attrs: Attributes, controller: js.Dynamic): Unit = {
    ...
  }
}

object TodoItemDirective {

  @JSExportAll
  @ExportToScope("directive")
  class Ctrl($scope: js.Dynamic) extends Controller {

    def onEditStart(): Unit = {
      $scope.editing = true
      $scope.title = $scope.todo.title
    }
 ...
  }
}
```

But even in this case, due to how the controller attribute in in the directive definition is used, the underlying javascript object for the fourth argument is just an empty object. 

I was looking at the js code generated by the compiler (it is not the code for the todo example, but for another directive):

``` js
ScalaJS.c.Lwebapp_App$$anonfun$main$5.prototype.apply__sjs_js_Object = (function() {
  var dimpl = new ScalaJS.c.Lwebapp_directives_CellText().init___();
  var ddo = {
    "templateUrl": dimpl.templateUrl$1,
    "scope": dimpl.isolateScope$1,
    "controllerAs": dimpl.controllerAs$1,
    "link": (function(dimpl$1) {
      return (function(scope$2, elem$2, attrs$2, controller$2) {
        var scope = scope$2;
        var elem = elem$2;
        var attrs = attrs$2;
        var controller = controller$2;
        dimpl$1.postLink__Lbiz_enef_angular_Scope__Lbiz_enef_angular_core_JQLite__Lbiz_enef_angular_core_Attributes__sjs_js_Dynamic__V(scope, elem, attrs, controller)
      })
    })(dimpl),
    "controller": ["$scope", "$timeout", "$rootScope", "$scope", (function(scope$2$1, x$macro$16$2, x$macro$17$2, x$macro$18$2) {
      var scope$1 = scope$2$1;
      var x$macro$16 = x$macro$16$2;
      var x$macro$17 = x$macro$17$2;
      var x$macro$18 = x$macro$18$2;
      var ctrl = new ScalaJS.c.Lwebapp_directives_CellText$CellTextController().init___Lbiz_enef_angular_core_Timeout__Lbiz_enef_angular_Scope__sjs_js_Dynamic(x$macro$16, x$macro$17, x$macro$18);
      scope$1["ctrl"] = ctrl
    })]
  };
  return ddo
});
```

If the controller function used `this`, as in:

``` js
controller: [ ...., function() {
  this.attr1 = ...
  this.attr2 = ...
}]
```

then angular would actually pass an instance of the controller as the fourth argument of the `postLink` function, and if the scalajs Directive trait allows to declare this fourth argument as an instance of the controller, everything would work the same as in angularjs (but type safe).

It seems to me that doing this would eliminate the need to use `@ExportToScope("ctrl")`.

Btw, the only way I found to use the `ctrl` attribute of the scope in the `postLink` function was to cast it to dynamic like this:

``` scala
val scp = scope.asInstanceOf[js.Dynamic]
scp.ctrl.myAttr....
```

**Comment by jokade** [comment]
First let me say that the directive api is the most experimental part of this lib, and it doesn't feel quite right to me yet (I actually don't like the `withController` and `@ExportToScope` semantics, but haven't found a better way to solve these problems). Also, there is currently no test suite for `Directive`, so it's probably buggy...

To answer your first question: you're right, the fourth argument to `postLink` should have the type of  `withController`. However, it could be useful if it was possible to declare any argument of the `postLink` function to be a `js.Dynamic` (e.g. a dynamic scope, allowing ad-hoc assignment of properties). What's your opinion on this?

About your second point: I didn't actually test using the `controller` argument passed to `postLink` -- and to be honest, I don't understand right now why I pass `$scope` as the fourth argument to the controller constructor array :(  Your suggestion looks good, but give me some time to check it out.

About `@ExportToScope`: We could export all (public) members of the directive controller to the scope, but then the user has no choice and it could be problematic (accidental shadowing...). WDYT?

**Comment by jokade** [comment]
@jpablo I've created a branch `issue28` with a test case for this issue (it's in `tests/src/test/scala/test/DirectiveTest`). You're right, the `controller` argument to the `postLink` function is an empty object, but I honestly don't know what the problem is (I rarely use controllers on directives) -- could you explain it to me?

About your suggestion to use `this` in the controller function: I don't think that it is possible to actually generate a [JS reference to `this`from Scala.js](http://www.scala-js.org/doc/calling-javascript.html); that's why the first parameter in the controller function is the `$scope`, which is used as `this`in this context.

**Comment by jpablo** [comment]
The controller argument is either a single object or a list of objects, depending on the `require` attribute. 

> controller - a controller instance - A controller instance if at least one directive on the element defines a > controller. The controller is shared among all the directives, which allows the directives to use the > controllers as a communication channel.

For example, this directive:

``` js
app.directive('exampleDirective', function() {
  return {
    scope: {},
    template: '<p>[{{ ctrl.x }}, {{ y }}, {{ z }}]</p>',
    controllerAs: 'ctrl', 
    controller: function($scope, $element){
      this.x = 1;
      $scope.y = 2;
    },
    link: function(scope, el, attr, controllerInstance) {
      scope.z = 3;
    }
  }
})
```

will print `[1, 2, 3]`.

So you can communicate with the view either augmenting the $scope (on the controller or on the link function) or by using `this` in the controller. Using `this` has some advantages because it maintains the separation between the angular $scope object ($on, $broadcast, $watch, etc) and the directive's own data (this.x, etc). 

The `$scope` argument still is required (besides using $on, etc) for communicating with other controllers using the `scope` description ( `scope: { attr1: '=', attr2: '@', ...}`).

Mhhh, not being able to het hold of `this` inside the controller function looks like a problem.

**Comment by jokade** [comment]
@jpablo Ok, when I first implemented the binding to Directive I actually didn't consider that controllers for directives are instantiated by Angular and just passed as `this` to the `controller(...)` function. Since more than one directive can extend this controller instance, I don't see an easy way to use Scala classes for directive controllers.

So I propose a complete re-design of the Directive API, as laid out in the example of issue #29, and I've implemented it already in branch `issue28`; if you want to test these changes, update your `build.sbt` with (still using Scala.js 0.6.0-RC2!): 

``` scala
resolvers += "Sonatype OSS Snapshots" at "https://oss.sonatype.org/content/repositories/snapshots"

libraryDependencies += "biz.enef" %%% "scalajs-angulate" % "issue28-SNAPSHOT"
```

I've also updated the [directives in the TodoMVC example](https://github.com/jokade/scalajs-angulate-todomvc/blob/v0.2/js/src/main/scala/todomvc/example/Directives.scala) to this new API.

I would be glad to hear your opinion on this!
